File: src/api/admin_dashboard.py
"""HOOPS AI - Admin Dashboard API"""
from datetime import date, timedelta, datetime
from fastapi import APIRouter, Depends
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from src.utils.database import get_db
from src.api.admin_auth import get_current_admin
from src.models.user import User
from src.models.team import Team
from src.models.team_member import TeamMember
from src.models.team_event import TeamEvent
from src.models.schedule_request import ScheduleRequest
from src.models.game_report import GameReport
from src.models.attendance import Attendance
from src.models.coach import Coach

router = APIRouter(prefix="/api/admin", tags=["admin-dashboard"])


@router.get("/dashboard")
async def admin_dashboard(admin: User = Depends(get_current_admin), db: AsyncSession = Depends(get_db)):
    """Rich aggregated data for admin dashboard."""

    # --- Get admin's teams ---
    teams_result = await db.execute(
        select(Team).where(Team.created_by_admin_id == admin.id, Team.is_active == True)
    )
    admin_teams = teams_result.scalars().all()
    admin_team_ids = [t.id for t in admin_teams]
    team_names = {t.id: t.name for t in admin_teams}

    empty = {
        "total_teams": 0, "total_coaches": 0, "total_players": 0, "total_parents": 0,
        "todays_events": [], "upcoming_events": [], "pending_requests": [],
        "recent_games": [], "new_members": [], "attendance_by_team": [],
    }
    if not admin_team_ids:
        return {"success": True, "data": empty}

    # --- Member counts ---
    counts_result = await db.execute(
        select(TeamMember.role_in_team, func.count(TeamMember.id))
        .where(TeamMember.team_id.in_(admin_team_ids), TeamMember.is_active == True)
        .group_by(TeamMember.role_in_team)
    )
    role_counts = {row[0]: row[1] for row in counts_result.all()}

    # --- Coach mapping (user_id → team_id, coach.id → team_id) ---
    coach_members_result = await db.execute(
        select(TeamMember.user_id, TeamMember.team_id)
        .where(
            TeamMember.team_id.in_(admin_team_ids),
            TeamMember.role_in_team == "coach",
            TeamMember.is_active == True,
        )
    )
    coach_user_to_team = {row[0]: row[1] for row in coach_members_result.all()}

    coach_id_to_team = {}
    coach_id_to_name = {}
    if coach_user_to_team:
        coaches_result = await db.execute(
            select(Coach).where(Coach.user_id.in_(list(coach_user_to_team.keys())))
        )
        for c in coaches_result.scalars().all():
            if c.user_id in coach_user_to_team:
                coach_id_to_team[c.id] = coach_user_to_team[c.user_id]
                coach_id_to_name[c.id] = c.name

    coach_ids = list(coach_id_to_team.keys())

    today = date.today()
    week_ahead = today + timedelta(days=7)
    week_ago = datetime.utcnow() - timedelta(days=7)

    # --- Today's events ---
    todays_result = await db.execute(
        select(TeamEvent)
        .where(TeamEvent.team_id.in_(admin_team_ids), TeamEvent.date == today, TeamEvent.is_active == True)
        .order_by(TeamEvent.time_start)
    )
    todays_events = [{
        "id": e.id, "title": e.title, "event_type": e.event_type,
        "time_start": e.time_start, "time_end": e.time_end,
        "location": e.location, "team_name": team_names.get(e.team_id, ""),
        "opponent": e.opponent,
    } for e in todays_result.scalars().all()]

    # --- Upcoming events (next 7 days, excluding today) ---
    upcoming_result = await db.execute(
        select(TeamEvent)
        .where(
            TeamEvent.team_id.in_(admin_team_ids),
            TeamEvent.date > today,
            TeamEvent.date <= week_ahead,
            TeamEvent.is_active == True,
        )
        .order_by(TeamEvent.date, TeamEvent.time_start)
        .limit(10)
    )
    upcoming_events = [{
        "id": e.id, "title": e.title, "event_type": e.event_type,
        "date": e.date.isoformat(), "time_start": e.time_start, "time_end": e.time_end,
        "location": e.location, "team_name": team_names.get(e.team_id, ""),
        "opponent": e.opponent,
    } for e in upcoming_result.scalars().all()]

    # --- Pending schedule requests ---
    pending_result = await db.execute(
        select(ScheduleRequest, Coach.name.label("coach_name"))
        .join(Coach, ScheduleRequest.coach_id == Coach.id)
        .where(ScheduleRequest.team_id.in_(admin_team_ids), ScheduleRequest.status == "pending")
        .order_by(ScheduleRequest.created_at.desc())
    )
    pending_requests = [{
        "id": row[0].id, "title": row[0].title, "event_type": row[0].event_type,
        "date": row[0].date.isoformat(), "time_start": row[0].time_start,
        "coach_name": row[1], "team_name": team_names.get(row[0].team_id, ""),
    } for row in pending_result.all()]

    # --- Recent game reports ---
    recent_games = []
    if coach_ids:
        games_result = await db.execute(
            select(GameReport)
            .where(GameReport.coach_id.in_(coach_ids))
            .order_by(GameReport.date.desc())
            .limit(5)
        )
        for g in games_result.scalars().all():
            team_id = coach_id_to_team.get(g.coach_id)
            recent_games.append({
                "date": g.date.isoformat(), "opponent": g.opponent,
                "result": g.result, "score_us": g.score_us, "score_them": g.score_them,
                "team_name": team_names.get(team_id, "") if team_id else "",
            })

    # --- New members (last 7 days) ---
    new_members_result = await db.execute(
        select(TeamMember, User)
        .join(User, TeamMember.user_id == User.id)
        .where(
            TeamMember.team_id.in_(admin_team_ids),
            TeamMember.joined_at >= week_ago,
            TeamMember.is_active == True,
        )
        .order_by(TeamMember.joined_at.desc())
        .limit(10)
    )
    new_members = [{
        "name": row[1].name, "role": row[0].role_in_team,
        "team_name": team_names.get(row[0].team_id, ""),
        "joined_at": row[0].joined_at.isoformat() if row[0].joined_at else "",
    } for row in new_members_result.all()]

    # --- Attendance by team ---
    attendance_by_team = []
    if coach_ids:
        att_total_result = await db.execute(
            select(Attendance.coach_id, func.count(Attendance.id).label("total"))
            .where(Attendance.coach_id.in_(coach_ids))
            .group_by(Attendance.coach_id)
        )
        att_present_result = await db.execute(
            select(Attendance.coach_id, func.count(Attendance.id).label("present"))
            .where(Attendance.coach_id.in_(coach_ids), Attendance.present == True)
            .group_by(Attendance.coach_id)
        )
        total_by_coach = {row[0]: row[1] for row in att_total_result.all()}
        present_by_coach = {row[0]: row[1] for row in att_present_result.all()}

        # Aggregate per team
        team_att = {}
        for cid, tid in coach_id_to_team.items():
            if cid not in total_by_coach:
                continue
            if tid not in team_att:
                team_att[tid] = {"total": 0, "present": 0}
            team_att[tid]["total"] += total_by_coach.get(cid, 0)
            team_att[tid]["present"] += present_by_coach.get(cid, 0)

        for tid, att in team_att.items():
            rate = round((att["present"] / att["total"] * 100) if att["total"] > 0 else 0)
            attendance_by_team.append({
                "team_name": team_names.get(tid, ""),
                "total": att["total"], "present": att["present"], "rate": rate,
            })
        attendance_by_team.sort(key=lambda x: x["rate"], reverse=True)

    return {
        "success": True,
        "data": {
            "total_teams": len(admin_teams),
            "total_coaches": role_counts.get("coach", 0),
            "total_players": role_counts.get("player", 0),
            "total_parents": role_counts.get("parent", 0),
            "todays_events": todays_events,
            "upcoming_events": upcoming_events,
            "pending_requests": pending_requests,
            "recent_games": recent_games,
            "new_members": new_members,
            "attendance_by_team": attendance_by_team,
        },
    }

File: static/js/admin_dashboard.js
/**
 * HOOPS AI — Admin Dashboard JS
 * Rich dashboard with data from across the system
 */

const EVENT_TYPE_HE = {
  practice: 'אימון', game: 'משחק', meeting: 'ישיבה', tournament: 'טורניר',
  team_building: 'גיבוש', team_dinner: 'ארוחת צוות', social: 'חברתי',
  tactical_video: 'וידאו טקטי', other: 'אחר',
};
const RESULT_MAP = {
  win:  { text: 'ניצחון', cls: 'result-win' },
  loss: { text: 'הפסד',  cls: 'result-loss' },
  draw: { text: 'תיקו',  cls: 'result-draw' },
};
const ROLE_HE = { coach: 'מאמן', player: 'שחקן', parent: 'הורה' };
const ROLE_ICON = { coach: 'sports', player: 'person', parent: 'family_restroom' };

document.addEventListener('DOMContentLoaded', () => {
  if (!AdminAPI.token) return;
  loadDashboard();
});

async function loadDashboard() {
  try {
    const res = await AdminAPI.get('/api/admin/dashboard');
    const d = res.data;

    // Stats cards
    document.getElementById('statTeams').textContent = d.total_teams;
    document.getElementById('statCoaches').textContent = d.total_coaches;
    document.getElementById('statPlayers').textContent = d.total_players;
    document.getElementById('statParents').textContent = d.total_parents;

    // Sections
    renderTodaysEvents(d.todays_events);
    renderPendingRequests(d.pending_requests);
    renderUpcomingEvents(d.upcoming_events);
    renderRecentGames(d.recent_games);
    renderAttendance(d.attendance_by_team);
    renderNewMembers(d.new_members);
  } catch (err) {
    console.error('Dashboard load error:', err);
  }
}

/* ══════════════════════════════════════
   Today's Activities
   ══════════════════════════════════════ */
function renderTodaysEvents(events) {
  const el = document.getElementById('todaysEvents');
  const badge = document.getElementById('todayCount');
  if (!events || !events.length) {
    el.innerHTML = `<div class="dash-empty">${t('admin.dashboard.empty.today')}</div>`;
    return;
  }
  badge.textContent = events.length;
  badge.style.display = '';
  const show = events.slice(0, 4);
  const more = events.length - show.length;
  el.innerHTML = show.map(e => `
    <div class="dash-event-item">
      <div class="dash-event-time">${e.time_start || '--:--'}${e.time_end ? ' - ' + e.time_end : ''}</div>
      <div class="dash-event-body">
        <div class="dash-event-title">
          <span class="schedule-type ${e.event_type}">${EVENT_TYPE_HE[e.event_type] || e.event_type}</span>
          ${esc(e.title)}
        </div>
        <div class="dash-event-meta">
          ${e.team_name ? '<span class="dash-team-tag">' + esc(e.team_name) + '</span>' : ''}
          ${e.location ? '<span class="material-symbols-outlined" style="font-size:13px;">location_on</span> ' + esc(e.location) : ''}
          ${e.opponent ? ' · vs ' + esc(e.opponent) : ''}
        </div>
      </div>
    </div>
  `).join('') + (more > 0 ? `<a href="/admin/schedule" class="dash-more-link">+${more} ${t('admin.dashboard.more_calendar')} →</a>` : '');
}

/* ══════════════════════════════════════
   Pending Approvals
   ══════════════════════════════════════ */
function renderPendingRequests(requests) {
  const el = document.getElementById('pendingRequests');
  const badge = document.getElementById('pendingBadge');
  if (!requests || !requests.length) {
    el.innerHTML = `<div class="dash-empty">${t('admin.dashboard.empty.pending')}</div>`;
    return;
  }
  badge.textContent = requests.length;
  badge.style.display = '';
  el.innerHTML = requests.map(r => `
    <div class="dash-request-item" id="request-${r.id}">
      <div class="dash-request-body">
        <div class="dash-event-title">
          <span class="schedule-type ${r.event_type}">${EVENT_TYPE_HE[r.event_type] || r.event_type}</span>
          ${esc(r.title)}
        </div>
        <div class="dash-event-meta">
          <span>${esc(r.coach_name)}</span>
          <span class="dash-team-tag">${esc(r.team_name)}</span>
          <span>${formatDateHe(r.date)}${r.time_start ? ' ' + r.time_start : ''}</span>
        </div>
      </div>
      <div class="dash-request-actions">
        <button class="dash-action-btn approve" onclick="quickApprove(${r.id})" title="${t('admin.dashboard.approve_title')}">
          <span class="material-symbols-outlined">check</span>
        </button>
        <button class="dash-action-btn reject" onclick="quickReject(${r.id})" title="${t('admin.dashboard.reject_title')}">
          <span class="material-symbols-outlined">close</span>
        </button>
      </div>
    </div>
  `).join('');
}

async function quickApprove(id) {
  try {
    await AdminAPI.put(`/api/schedule-requests/${id}/approve`, {});
    document.getElementById(`request-${id}`)?.remove();
    AdminToast.success(t('admin.dashboard.request_approved'));
    _updatePendingCount();
  } catch { AdminToast.error(t('admin.dashboard.approve_error')); }
}

async function quickReject(id) {
  try {
    await AdminAPI.put(`/api/schedule-requests/${id}/reject`, {});
    document.getElementById(`request-${id}`)?.remove();
    AdminToast.success(t('admin.dashboard.request_rejected'));
    _updatePendingCount();
  } catch { AdminToast.error(t('admin.dashboard.reject_error')); }
}

function _updatePendingCount() {
  const badge = document.getElementById('pendingBadge');
  const container = document.getElementById('pendingRequests');
  const remaining = container.querySelectorAll('.dash-request-item').length;
  if (remaining === 0) {
    badge.style.display = 'none';
    container.innerHTML = `<div class="dash-empty">${t('admin.dashboard.empty.pending')}</div>`;
  } else {
    badge.textContent = remaining;
  }
}

/* ══════════════════════════════════════
   Upcoming Events (7 days)
   ══════════════════════════════════════ */
function renderUpcomingEvents(events) {
  const el = document.getElementById('upcomingEvents');
  if (!events || !events.length) {
    el.innerHTML = `<div class="dash-empty">${t('admin.dashboard.empty.upcoming')}</div>`;
    return;
  }
  const show = events.slice(0, 4);
  const more = events.length - show.length;
  el.innerHTML = show.map(e => {
    const d = new Date(e.date);
    return `
      <div class="dash-event-item">
        <div class="dash-event-date-col">
          <div class="dash-date-day">${d.getDate()}</div>
          <div class="dash-date-month">${d.toLocaleDateString('he-IL', { month: 'short' })}</div>
          <div class="dash-date-dow">${d.toLocaleDateString('he-IL', { weekday: 'short' })}</div>
        </div>
        <div class="dash-event-body">
          <div class="dash-event-title">
            <span class="schedule-type ${e.event_type}">${EVENT_TYPE_HE[e.event_type] || e.event_type}</span>
            ${esc(e.title)}
          </div>
          <div class="dash-event-meta">
            ${e.team_name ? '<span class="dash-team-tag">' + esc(e.team_name) + '</span>' : ''}
            ${e.time_start ? e.time_start : ''}
            ${e.location ? ' · ' + esc(e.location) : ''}
            ${e.opponent ? ' · vs ' + esc(e.opponent) : ''}
          </div>
        </div>
      </div>
    `;
  }).join('') + (more > 0 ? `<a href="/admin/schedule" class="dash-more-link">+${more} ${t('admin.dashboard.more_calendar')} →</a>` : '');
}

/* ══════════════════════════════════════
   Recent Game Results
   ══════════════════════════════════════ */
function renderRecentGames(games) {
  const el = document.getElementById('recentGames');
  if (!games || !games.length) {
    el.innerHTML = `<div class="dash-empty">${t('admin.dashboard.empty.games')}</div>`;
    return;
  }
  el.innerHTML = games.map(g => {
    const r = RESULT_MAP[g.result] || { text: g.result, cls: '' };
    return `
      <div class="dash-game-item">
        <div class="dash-game-result ${r.cls}">${r.text}</div>
        <div class="dash-game-body">
          <div class="dash-game-score">${g.score_us != null ? '<span dir="ltr">' + g.score_us + ' - ' + g.score_them + '</span>' : t('admin.dashboard.score_pending')}</div>
          <div class="dash-event-meta">
            vs ${esc(g.opponent)}
            ${g.team_name ? ' · <span class="dash-team-tag">' + esc(g.team_name) + '</span>' : ''}
          </div>
        </div>
        <div class="dash-game-date">${formatDateHe(g.date)}</div>
      </div>
    `;
  }).join('');
}

/* ══════════════════════════════════════
   Attendance Overview
   ══════════════════════════════════════ */
function renderAttendance(teams) {
  const el = document.getElementById('attendanceSummary');
  if (!teams || !teams.length) {
    el.innerHTML = `<div class="dash-empty">${t('admin.dashboard.empty.attendance')}</div>`;
    return;
  }
  el.innerHTML = teams.map(tm => {
    const lvl = tm.rate >= 80 ? 'good' : tm.rate >= 60 ? 'avg' : 'low';
    return `
      <div class="dash-att-item">
        <div class="dash-att-header">
          <span class="dash-team-tag">${esc(tm.team_name)}</span>
          <span class="dash-att-rate ${lvl}">${tm.rate}%</span>
        </div>
        <div class="dash-att-bar"><div class="dash-att-fill ${lvl}" style="width:${tm.rate}%"></div></div>
        <div class="dash-att-detail">${tm.present} / ${tm.total} ${t('admin.dashboard.attendance_records')}</div>
      </div>
    `;
  }).join('');
}

/* ══════════════════════════════════════
   New Members
   ══════════════════════════════════════ */
function renderNewMembers(members) {
  const el = document.getElementById('newMembers');
  if (!members || !members.length) {
    el.innerHTML = `<div class="dash-empty">${t('admin.dashboard.empty.new_members')}</div>`;
    return;
  }
  el.innerHTML = members.map(m => `
    <div class="dash-member-item">
      <div class="dash-member-avatar">
        <span class="material-symbols-outlined">${ROLE_ICON[m.role] || 'person'}</span>
      </div>
      <div class="dash-member-body">
        <div class="dash-member-name">${esc(m.name)}</div>
        <div class="dash-event-meta">
          <span class="dash-role-chip ${m.role}">${ROLE_HE[m.role] || m.role}</span>
          <span class="dash-team-tag">${esc(m.team_name)}</span>
        </div>
      </div>
      <div class="dash-member-time">${timeAgoShort(m.joined_at)}</div>
    </div>
  `).join('');
}

/* ══════════════════════════════════════
   Helpers
   ══════════════════════════════════════ */
/* formatDateHe → shared-utils.js */

function timeAgoShort(dateStr) {
  if (!dateStr) return '';
  const d = new Date(dateStr.endsWith('Z') ? dateStr : dateStr + 'Z');
  const diff = Date.now() - d.getTime();
  const days = Math.floor(diff / 86400000);
  if (days === 0) return t('admin.dashboard.today');
  if (days === 1) return t('admin.dashboard.yesterday');
  return t('admin.dashboard.days_ago', { count: days });
}
